### PR TITLE
Category type.

### PR DESCRIPTION
**Pull request by 2e12**
Create new Category type. A Categroy type is only addable to a folder.

Bildschirmfoto 2020-04-01 um 14 54 13

Bildschirmfoto 2020-04-01 um 14 53 57

### CODE REVIEW

File: ftw/noticeboard/configure.zcml [hunk 1]
```diff
@@ -28,4 +29,8 @@
         directory="upgrades"
         />
 
+    <include file="permission.zcml" />
+    <include zcml:condition="installed ftw.lawgiver" file="lawgiver.zcml" />
```
Comment (mbaechtold) [inline ftw/noticeboard/configure.zcml:40]: Was wird damit bezweckt?

File: ftw/noticeboard/profiles/default/types.xml [hunk 2]
```diff
@@ -0,0 +1,5 @@
+<?xml version="1.0"?>
+<object name="portal_types" meta_type="Plone Types Tool">
+    <object name="ftw.noticeboard.Category" meta_type="Dexterity FTI"/>
+    <object name="Folder" meta_type="Dexterity FTI"/>
```
Comment (mbaechtold) [inline ftw/noticeboard/profiles/default/types.xml]: ✂️ Im `types.xml` werden nur die Typen eingetragen, die das Addon selber mitbringt.

File: ftw/noticeboard/types/category.py [hunk 3]
```diff
@@ -0,0 +1,23 @@
+from ftw.noticeboard.interfaces import ICategory
+from plone.app.dexterity.behaviors.metadata import IBasic
+from plone.dexterity.content import Container
+from plone.supermodel import model
+from zope.interface import alsoProvides
+from zope.interface import implementer
+from zope.interface import implements
+from zope.schema.interfaces import ICollection
+
```
Comment (mbaechtold) [inline ftw/noticeboard/types/category.py:9]: Is the number of lines compatible with PEP-8?

File: test-plone-4.3.x.cfg [hunk 4]
```diff
@@ -9,5 +9,5 @@ versions = versions
 
 
 [versions]
-setuptools = 28.2.0
-zc.buildout = 2.11.0
+setuptools = 41.6.0
```
Comment (mbaechtold) [inline test-plone-4.3.x.cfg]: Please move that to a separate pull request where you can explain in depth why this is necessary 😉

**Review by mbaechtold — CHANGES_REQUESTED** [review]

File: ftw/noticeboard/configure.zcml (hunk 1 above)
Comment (2e12) [reply ftw/noticeboard/configure.zcml:40]: Damit wird alles was mit den Berechtigungen/ftw.lawgiver in eigene Dateien ausgelagert. Das wird so in vielen ftw.* Packages gemacht. So zum Beispiel auch bei ftw.news -> https://github.com/4teamwork/ftw.news/blob/master/ftw/news/configure.zcml#L18-L19

Comment (mbaechtold) [reply ftw/noticeboard/configure.zcml:40]: Sehr verständlich, danke für die Ausführung!

File: ftw/noticeboard/tests/test_categroy_permission.py [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+        # Try to add a new Category
+        try:
+            # When the expectation 'can add' is, then this should work fine.
+            # When the expectation 'can not add' is, then this should fail.
+            factoriesmenu.add('Category')
+            if expectation == 'can not add':
+                self.fail('The role {} shouldn\'t be able to add a new Category type.'.format(role))
+        except ValueError as e:
+            if 'The type "Category" is not addable' in e.message and expectation == 'can add':
+                self.fail('The role {} should be able to add a new Category type.'.format(role))
```
Comment (mbaechtold) [inline ftw/noticeboard/tests/test_categroy_permission.py:31]: Just a comment: I really had focus in order to understand this code.

File: ftw/noticeboard/types/category.py [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+    """
+    pass
+
+
+alsoProvides(ICategorySchema, IBasic)
+
+
+class Category(Container):
+    implements(ICategory)
+
```
Comment (mbaechtold) [inline ftw/noticeboard/types/category.py]: ✂️You may configure your editor to automatically remove empty lines at the end of the file.

File: versions.cfg [hunk 7]
```diff
@@ -1,6 +1,7 @@
 [versions]
-six = 1.12.0
+plone.directives.dexterity = 1.0.2
```
Comment (mbaechtold) [inline versions.cfg]: What do these version pinnings have to do with this commit which states that it will `Add category type.`? 

Please explain why these are needed in the commit message. You may want to move these changes to a separate commit where you can use its message body to elaborate the reason for the change.

**Review by mbaechtold — CHANGES_REQUESTED** [review]

File: versions.cfg (hunk 7 above)
Comment (2e12) [reply versions.cfg]: Checked the changes and determined that they are not necessary. Changes are removed now.

File: ftw/noticeboard/types/category.py [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+from plone.app.dexterity.behaviors.metadata import IBasic
+from plone.dexterity.content import Container
+from plone.supermodel import model
+from zope.interface import alsoProvides
+from zope.interface import implementer
+from zope.interface import implements
+from zope.schema.interfaces import ICollection
+
+
+@implementer(ICollection)
```
Comment (mbaechtold) [inline ftw/noticeboard/types/category.py:11]: What does this do?

File: ftw/noticeboard/types/category.py [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+class ICategorySchema(model.Schema):
+    """
+    This schema represents a category item.
+    """
+    pass
+
+
+alsoProvides(ICategorySchema, IBasic)
```
Comment (mbaechtold) [inline ftw/noticeboard/types/category.py:19]: What does this do?

**Review by mbaechtold — CHANGES_REQUESTED** [review]

File: ftw/noticeboard/types/category.py (hunk 9 above)
Comment (2e12) [reply ftw/noticeboard/types/category.py:19]: This implements the IBasic behaviour.

File: ftw/noticeboard/types/category.py (hunk 8 above)
Comment (2e12) [reply ftw/noticeboard/types/category.py:11]: This does implement the ICollection behaviour.

**Comment by 2e12** [comment]
@mbaechtold Should I implement those two behaviour the same way?

File: ftw/noticeboard/types/category.py (hunk 8 above)
Comment (mbaechtold) [reply ftw/noticeboard/types/category.py:11]: Why would we want that?

File: ftw/noticeboard/types/category.py (hunk 9 above)
Comment (mbaechtold) [reply ftw/noticeboard/types/category.py:19]: But the behavior is added in the FTI (https://github.com/4teamwork/ftw.noticeboard/pull/2/files#diff-a1ca04c099d686312c1c53739bb70760R21) so this is not needed. We could provide the FormFields instead. See https://github.com/4teamwork/ftw.news/blob/cf2a4ecda4dc5f14f93c002d95b52d5b90ac1437/ftw/news/contents/news_folder.py#L16

**Review by mbaechtold — CHANGES_REQUESTED** [review]
Let's try to finish this pull request by adding the workflow as required in the issue (see https://github.com/4teamwork/bern.web/issues/1604). Among other, this is what's needed:
- A `workflows.xml` in the profile
- A folder in the profile containing a `specification.txt` and `definition.xml`

File: ftw/noticeboard/profiles/example_workflow/workflows.xml [hunk 10]
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0"?>
+<object name="portal_workflow" meta_type="Plone Workflow Tool">
+    <object name="noticeboard_notice_workflow" meta_type="Workflow"/>
```
Comment (mbaechtold) [inline ftw/noticeboard/profiles/example_workflow/workflows.xml]: In diesem Pull Request geht es um Category, nicht um Notice, oder? 
noticeboard_notice_workflow -> noticeboard_category_workflow

File: ftw/noticeboard/profiles/example_workflow/workflows.xml [hunk 11]
```diff
@@ -0,0 +1,9 @@
...
+    <bindings>
+        <type type_id="ftw.noticeboard.Category">
+            <bound-workflow workflow_id="noticeboard_notice_workflow"/>
```
Comment (mbaechtold) [inline ftw/noticeboard/profiles/example_workflow/workflows.xml]: In diesem Pull Request geht es um Category, nicht um Notice, oder? 
noticeboard_notice_workflow -> noticeboard_category_workflow

**Review by mbaechtold — CHANGES_REQUESTED** [review]
Bitte generiere noch das `definition.xml` mit Hilfe von `ftw.lawgiver`.

**Comment by 2e12** [comment]
@mbaechtold  So, habe das was ich konnte korrigiert. So wie es aussieht wird mein Workflow nicht erkennt... Also es ist mir nicht möglich eine definition zu definieren. Ich habe alles noch ein mal mit dem workflow aus dem Notice pr abgeglichen und kann meinen Fehler nicht finden ...

**Comment by mbaechtold** [comment]
Siehe https://docs.plone.org/develop/addons/components/genericsetup.html#workflows-xml

> Next to this, the workflows directory is checked. This contains sub directories with the same name as the workflows. Each sub directory contains a file definition.xml with the definition for this workflow.

Der Ordner mit dem Workflow muss in einem Unterordner `workflows` platziert werden, also unter `ftw/noticeboard/profiles/example_workflow/workflows/noticeboard_category_workflow`.

File: ftw/noticeboard/types/category.py (hunk 8 above)
Comment (mbaechtold) [reply ftw/noticeboard/types/category.py:11]: @2e12 ❓

File: ftw/noticeboard/types/category.py (hunk 9 above)
Comment (mbaechtold) [reply ftw/noticeboard/types/category.py:19]: @2e12, any objection?

**Comment by 2e12** [comment]
@mbaechtold Jetzt geht es, mercii :-)
